test/custom-shallow-renderer/utils.ts

```diff
@@ -0,0 +1,116 @@
+import { options } from 'preact';
+import type { VNode } from 'preact';
+import sinon from 'sinon';
+
+// These tests are from react-shallow-renderer which uses Jest for testing, so
+// here is a lightweight adapter around `chai` to give it the same API as Jest's
+// `expect` so we don't have to rewrite all the `expect` statements to `assert`
+function createMatcher<T>(actual: T, negate = false) {
+  return {
+    toBe(expected: T) {
+      if (negate) {
+        assert.notEqual(actual, expected);
+      } else {
+        assert.equal(actual, expected);
+      }
+    },
+    toEqual(expected: T) {
+      if (negate) {
+        assert.notDeepEqual(actual, expected);
+      } else {
+        assert.deepEqual(actual, expected);
+      }
+    },
+    toBeNull() {
+      if (negate) {
+        assert.isNotNull(actual);
+      } else {
+        assert.isNull(actual);
+      }
+    },
+    toThrowError(expected: string) {
+      if (negate) {
+        assert.doesNotThrow(actual as any, expected);
+      } else {
+        assert.throws(actual as any, expected);
+      }
+    },
+    toErrorDev(expected: any, options?: any) {
+      if (negate) {
+        assert.doesNotThrow(actual as any, expected);
+      } else {
+        assert.throws(actual as any, expected);
+      }
+    },
+    toBeCalled() {
+      if (negate) {
+        sinon.assert.notCalled(actual as any);
+      } else {
+        sinon.assert.called(actual as any);
+      }
+    },
+    toHaveBeenCalled() {
+      if (negate) {
+        sinon.assert.notCalled(actual as any);
+      } else {
+        sinon.assert.called(actual as any);
+      }
+    },
+    toHaveBeenCalledTimes(count: number) {
+      if (negate) {
+        throw new Error(`.not.toHaveBeenCalledTimes is not supported`);
+      } else {
+        sinon.assert.callCount(actual as any, count);
+      }
+    },
+  };
+}
+
+type Matchers = ReturnType<typeof createMatcher>;
+type ExpectMatchers = Matchers & { not: Matchers };
+
+export function expect<T>(actual: T): ExpectMatchers {
+  const matchers: ExpectMatchers = createMatcher(actual) as any;
+  matchers.not = createMatcher(actual, true);
+
+  return matchers;
+}
+
+export function installVNodeTestHook() {
+  let prevVNodeHook: ((vnode: VNode) => void) | undefined;
+
+  before(() => {
+    prevVNodeHook = options.vnode;
+    options.vnode = vnode => {
+      if (prevVNodeHook) {
+        prevVNodeHook(vnode);
+      }
+
+      // Override the vnodeId (`__v`) to NaN so that we can compare VNodes in
+      // tests and verify expected VNode output. We choose NaN here because it
+      // successfully threads the desired behavior we want for these tests.
+      //
+      // The Preact diff uses `__v` to shortcut diffing VNodes that haven't
+      // changed since creation (it treats VNodes as immutable). So when Preact
+      // checks if `oldVNode.__v == newVNode.__v`, setting `__v` to NaN will
+      // always return false, per the rules of JavaScript (NaN !== NaN in JS).
```

Comment:
I presume React doesn't have this short-circuiting behavior?

Reply:
I'm not sure. It might have something similar but I think it does it it in a different way that doesn't rely on private VNode properties (it has been a while since I looked).